File: control-plane/frontend/src/utils/resourceValidation.ts
// Validation helpers for Kubernetes-style resource quantity strings used by
// instance and global-default settings. Mirrors the backend rules in
// control-plane/internal/handlers/resourcevalidation.go.

const CPU_RE = /^(\d+m|\d+(\.\d+)?)$/;
const MEM_RE = /^\d+(Ki|Mi|Gi)$/;

export const isValidCPU = (v: string): boolean => CPU_RE.test(v);
export const isValidMemory = (v: string): boolean => MEM_RE.test(v);
export const isValidStorage = isValidMemory;
export const isValidResolution = (v: string): boolean => v === "" || /^\d+x\d+$/.test(v);

export const cpuToMillis = (v: string): number => {
  if (v.endsWith("m")) return parseInt(v, 10);
  return parseFloat(v) * 1000;
};

export const memToBytes = (v: string): number => {
  const n = parseInt(v, 10);
  if (v.endsWith("Gi")) return n * 1024 * 1024 * 1024;
  if (v.endsWith("Mi")) return n * 1024 * 1024;
  if (v.endsWith("Ki")) return n * 1024;
  return n;
};

export type ResourceQuantities = {
  cpu_request?: string;
  cpu_limit?: string;
  memory_request?: string;
  memory_limit?: string;
  storage_home?: string;
  storage_homebrew?: string;
};

// Returns a per-field error map. Empty values are treated as "not provided"
// and skipped, matching the backend tolerance.
export function validateResourceQuantities(
  q: ResourceQuantities,
): Partial<Record<keyof ResourceQuantities | "cpu_pair" | "memory_pair", string>> {
  const errors: Partial<
    Record<keyof ResourceQuantities | "cpu_pair" | "memory_pair", string>
  > = {};

  if (q.cpu_request && !isValidCPU(q.cpu_request))
    errors.cpu_request = "e.g. 500m or 0.5";
  if (q.cpu_limit && !isValidCPU(q.cpu_limit))
    errors.cpu_limit = "e.g. 2000m or 2";
  if (q.memory_request && !isValidMemory(q.memory_request))
    errors.memory_request = "e.g. 1Gi or 512Mi";
  if (q.memory_limit && !isValidMemory(q.memory_limit))
    errors.memory_limit = "e.g. 4Gi or 2048Mi";
  if (q.storage_home && !isValidStorage(q.storage_home))
    errors.storage_home = "e.g. 10Gi";
  if (q.storage_homebrew && !isValidStorage(q.storage_homebrew))
    errors.storage_homebrew = "e.g. 10Gi";

  if (
    !errors.cpu_request &&
    !errors.cpu_limit &&
    q.cpu_request &&
    q.cpu_limit &&
    cpuToMillis(q.cpu_request) > cpuToMillis(q.cpu_limit)
  ) {
    errors.cpu_pair = "CPU request cannot exceed CPU limit";
  }
  if (
    !errors.memory_request &&
    !errors.memory_limit &&
    q.memory_request &&
    q.memory_limit &&
    memToBytes(q.memory_request) > memToBytes(q.memory_limit)
  ) {
    errors.memory_pair = "Memory request cannot exceed memory limit";
  }
  return errors;
}

File: control-plane/internal/handlers/resourcevalidation.go
package handlers

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var (
	cpuRegex        = regexp.MustCompile(`^(\d+m|\d+(\.\d+)?)$`)
	memoryRegex     = regexp.MustCompile(`^\d+(Ki|Mi|Gi)$`)
	resolutionRegex = regexp.MustCompile(`^\d+x\d+$`)
)

func cpuToMillicores(s string) int64 {
	if strings.HasSuffix(s, "m") {
		n, _ := strconv.ParseInt(s[:len(s)-1], 10, 64)
		return n
	}
	f, _ := strconv.ParseFloat(s, 64)
	return int64(f * 1000)
}

func memoryToBytes(s string) int64 {
	unitMap := map[string]int64{"Ki": 1024, "Mi": 1024 * 1024, "Gi": 1024 * 1024 * 1024}
	for suffix, multiplier := range unitMap {
		if strings.HasSuffix(s, suffix) {
			n, _ := strconv.ParseInt(s[:len(s)-len(suffix)], 10, 64)
			return n * multiplier
		}
	}
	n, _ := strconv.ParseInt(s, 10, 64)
	return n
}

// ResourceQuantities holds the six string values that get validated together.
// Empty strings are treated as "not provided" and skipped.
type ResourceQuantities struct {
	CPURequest      string
	CPULimit        string
	MemoryRequest   string
	MemoryLimit     string
	StorageHome     string
	StorageHomebrew string
}

// ValidateResourceQuantities checks each provided field against the expected
// k8s-quantity format and enforces request <= limit when both sides are
// present. It returns a user-facing error suitable for a 400 response.
func ValidateResourceQuantities(q ResourceQuantities) error {
	if q.CPURequest != "" && !cpuRegex.MatchString(q.CPURequest) {
		return fmt.Errorf("Invalid CPU request format (e.g., 500m or 0.5)")
	}
	if q.CPULimit != "" && !cpuRegex.MatchString(q.CPULimit) {
		return fmt.Errorf("Invalid CPU limit format (e.g., 2000m or 2)")
	}
	if q.MemoryRequest != "" && !memoryRegex.MatchString(q.MemoryRequest) {
		return fmt.Errorf("Invalid memory request format (e.g., 1Gi or 512Mi)")
	}
	if q.MemoryLimit != "" && !memoryRegex.MatchString(q.MemoryLimit) {
		return fmt.Errorf("Invalid memory limit format (e.g., 4Gi or 2048Mi)")
	}
	if q.StorageHome != "" && !memoryRegex.MatchString(q.StorageHome) {
		return fmt.Errorf("Invalid home storage format (e.g., 10Gi or 512Mi)")
	}
	if q.StorageHomebrew != "" && !memoryRegex.MatchString(q.StorageHomebrew) {
		return fmt.Errorf("Invalid homebrew storage format (e.g., 10Gi or 512Mi)")
	}

	if q.CPURequest != "" && q.CPULimit != "" &&
		cpuToMillicores(q.CPURequest) > cpuToMillicores(q.CPULimit) {
		return fmt.Errorf("CPU request cannot exceed CPU limit")
	}
	if q.MemoryRequest != "" && q.MemoryLimit != "" &&
		memoryToBytes(q.MemoryRequest) > memoryToBytes(q.MemoryLimit) {
		return fmt.Errorf("Memory request cannot exceed memory limit")
	}
	return nil
}
